sway-features/src/lib.rs

```diff
@@ -0,0 +1,231 @@
+use clap::{Parser, ValueEnum};
+
+macro_rules! features {
+    (count; ) => {0};
+    (count; $name:ident $($rest:tt)?) => {1 + features!{count; $($rest)*}};
+    ($($name:ident = $enabled:literal, $url:literal),* $(,)?) => {
+        paste::paste! {
+            pub const CFG: [&str; features!{count; $($name)*}] = [
+                $(
+                    stringify!([<experimental_ $name:snake>]),
+                )*
+            ];
+
+            #[derive(Copy, Clone, Debug, ValueEnum)]
+            #[value(rename_all = "snake")]
+            pub enum Features {
+                $(
+                    [<$name:camel>],
+                )*
+            }
+
+            impl std::str::FromStr for Features {
+                type Err = Error;
+
+                fn from_str(s: &str) -> Result<Self, Self::Err> {
+                    match s {
+                        $(
+                            stringify!([<$name:snake>]) => {
+                                Ok(Self::[<$name:camel>])
+                            },
+                        )*
+                        _ => Err(Error::UnknownFeature(s.to_string())),
+                    }
+                }
+            }
+
+            #[derive(Copy, Clone, Debug, PartialEq, Eq)]
+            pub struct ExperimentalFeatures {
+                $(
+                    pub [<$name:snake>]: bool,
+                )*
+            }
+
+            impl std::default::Default for ExperimentalFeatures {
+                fn default() -> Self {
+                    Self {
+                        $(
+                            [<$name:snake>]: $enabled,
+                        )*
+                    }
+                }
+            }
+
+            impl ExperimentalFeatures {
+                pub fn set_enabled(&mut self, feature: &str, enabled: bool) -> Result<(), Error> {
+                    let feature = feature.trim();
+                    match feature {
+                        $(
+                            stringify!([<$name:snake>]) => {
+                                self.[<$name:snake>] = enabled;
+                                Ok(())
+                            },
+                        )*
+                        "" => Ok(()),
+                        _ => Err(Error::UnknownFeature(feature.to_string())),
+                    }
+                }
+
+                pub fn enable_feature(&mut self, feature: Features, enabled: bool) {
```

Comment:
When seeing it for the first time in the usage code, the function name created a wrong expectation of doing enabling only. It was definitely confusing. Since it is. unlike `set_enabled`, always used with explicit `true` or `false` the proposal is to have two explicit methods `enable_feature` and `disable_feature`. It will also make callers code more readable and  explicit.

Reply:
I changed some of the names. Hope is better now.